File: orchestrator/resolver.go
package main

import (
	"database/sql"
	"fmt"
	"log"
	"strconv"
	"strings"

	_ "github.com/lib/pq"
)

// Candidate represents an entity match from the catalog.
type Candidate struct {
	ItemID   int64
	Kind     string
	Domain   string
	EntityID string
	Name     string
	Area     string
	Score    float64 // 1.0 = perfect match, 0.0 = no match
}

// Resolver handles entity matching against the pgvector catalog.
type Resolver struct {
	db       *sql.DB
	embedder *EmbedderClient
}

// NewResolver connects to PostgreSQL.
func NewResolver(dsn string, embedder *EmbedderClient) (*Resolver, error) {
	db, err := sql.Open("postgres", dsn)
	if err != nil {
		return nil, fmt.Errorf("pgvector connect: %w", err)
	}
	if err := db.Ping(); err != nil {
		return nil, fmt.Errorf("pgvector ping: %w", err)
	}
	db.SetMaxOpenConns(5)
	db.SetMaxIdleConns(2)
	log.Println("Connected to pgvector catalog")
	return &Resolver{db: db, embedder: embedder}, nil
}

// Close closes the DB connection.
func (r *Resolver) Close() {
	r.db.Close()
}

// Ping checks the DB connection.
func (r *Resolver) Ping() error {
	return r.db.Ping()
}

// Resolve finds matching entities for a user utterance.
// It runs lexical match first, then vector search, and merges results.
func (r *Resolver) Resolve(text string) ([]Candidate, error) {
	lower := strings.ToLower(strings.TrimSpace(text))

	// Pass A: Lexical matching
	lexical, err := r.lexicalMatch(lower)
	if err != nil {
		log.Printf("Lexical match error (non-fatal): %v", err)
	}
	lexical = dedupeCandidates(lexical)

	// If we got a strong lexical match (exact entity_id or exact name), return it
	for _, c := range lexical {
		if c.Score >= 0.95 {
			log.Printf("Strong lexical match: %s (%.2f)", c.EntityID, c.Score)
			return []Candidate{c}, nil
		}
	}

	// Pass B: Vector search via embedder + pgvector; if unavailable, fall back to text search.
	var vector []Candidate
	if r.embedder != nil {
		emb, embErr := r.embedder.Embed(lower)
		if embErr != nil {
			log.Printf("Embedder error (falling back to text search): %v", embErr)
		} else {
			vector, err = r.vectorSearch(emb)
			if err != nil {
				log.Printf("Vector search error (falling back to text search): %v", err)
				vector = nil
			}
		}
	}
	if len(vector) == 0 {
		vector, err = r.textSearch(lower)
		if err != nil {
			log.Printf("Text search error (non-fatal): %v", err)
		}
	}
	vector = dedupeCandidates(vector)

	// Merge results, lexical first (higher priority) and keep highest score per entity.
	byID := make(map[string]Candidate, len(lexical)+len(vector))
	order := make([]string, 0, len(lexical)+len(vector))
	for _, c := range lexical {
		if _, ok := byID[c.EntityID]; !ok {
			order = append(order, c.EntityID)
		}
		byID[c.EntityID] = c
	}
	for _, c := range vector {
		if existing, ok := byID[c.EntityID]; ok {
			if c.Score > existing.Score {
				byID[c.EntityID] = c
			}
		} else {
			order = append(order, c.EntityID)
			byID[c.EntityID] = c
		}
	}
	merged := make([]Candidate, 0, len(byID))
	for _, id := range order {
		if c, ok := byID[id]; ok {
			merged = append(merged, c)
		}
	}

	// Sort by score descending
	for i := 0; i < len(merged); i++ {
		for j := i + 1; j < len(merged); j++ {
			if merged[j].Score > merged[i].Score {
				merged[i], merged[j] = merged[j], merged[i]
			}
		}
	}

	// Limit to top 8
	if len(merged) > 8 {
		merged = merged[:8]
	}

	return merged, nil
}

func (r *Resolver) vectorSearch(embedding []float64) ([]Candidate, error) {
	if len(embedding) == 0 {
		return nil, nil
	}

	vectorLiteral := formatVector(embedding)
	rows, err := r.db.Query(`
		SELECT item_id, kind, domain, entity_id, name, COALESCE(area, ''),
		       1 - (embedding <=> $1::vector) AS score
		FROM catalog_items
		WHERE enabled = true
		ORDER BY embedding <=> $1::vector
		LIMIT 8
	`, vectorLiteral)
	if err != nil {
		return nil, fmt.Errorf("vector query: %w", err)
	}
	defer rows.Close()

	var candidates []Candidate
	for rows.Next() {
		var c Candidate
		if err := rows.Scan(&c.ItemID, &c.Kind, &c.Domain, &c.EntityID, &c.Name, &c.Area, &c.Score); err != nil {
			return nil, fmt.Errorf("vector scan: %w", err)
		}
		candidates = append(candidates, c)
	}
	return candidates, nil
}

func formatVector(v []float64) string {
	var b strings.Builder
	b.Grow(len(v) * 8)
	b.WriteByte('[')
	for i, n := range v {
		if i > 0 {
			b.WriteByte(',')
		}
		b.WriteString(strconv.FormatFloat(n, 'f', -1, 64))
	}
	b.WriteByte(']')
	return b.String()
}

// lexicalMatch performs exact and fuzzy text matching against the catalog.
func (r *Resolver) lexicalMatch(text string) ([]Candidate, error) {
	var candidates []Candidate

	// Exact entity_id match
	rows, err := r.db.Query(`
		SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
		FROM catalog_items
		WHERE enabled = true AND entity_id = $1
	`, text)
	if err == nil {
		defer rows.Close()
		for rows.Next() {
			var c Candidate
			rows.Scan(&c.ItemID, &c.Kind, &c.Domain, &c.EntityID, &c.Name, &c.Area)
			c.Score = 1.0
			candidates = append(candidates, c)
		}
	}

	// Name match (case-insensitive, contains)
	words := extractKeywords(text)
	if len(words) > 0 {
		// Build a LIKE pattern from important words
		for _, word := range words {
			if len(word) < 3 {
				continue
			}
			rows, err := r.db.Query(`
				SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
				FROM catalog_items
				WHERE enabled = true AND lower(name) LIKE '%' || $1 || '%'
				LIMIT 10
			`, word)
			if err != nil {
				continue
			}
			for rows.Next() {
				var c Candidate
				rows.Scan(&c.ItemID, &c.Kind, &c.Domain, &c.EntityID, &c.Name, &c.Area)
				// Score based on how well the name matches
				nameLower := strings.ToLower(c.Name)
				if nameLower == text {
					c.Score = 0.95
				} else if strings.Contains(nameLower, text) {
					c.Score = 0.85
				} else {
					c.Score = 0.6 + 0.1*float64(len(word))/float64(len(nameLower))
				}
				candidates = append(candidates, c)
			}
			rows.Close()
		}
	}

	return candidates, nil
}

// textSearch does a broader text-based search for entities.
func (r *Resolver) textSearch(text string) ([]Candidate, error) {
	var candidates []Candidate

	// Search by matching any of the significant words against name, domain, area
	words := extractKeywords(text)
	if len(words) == 0 {
		return nil, nil
	}

	// Build OR conditions for each word
	conditions := make([]string, 0, len(words))
	args := make([]interface{}, 0, len(words))
	for i, w := range words {
		if len(w) < 3 {
			continue
		}
		conditions = append(conditions, fmt.Sprintf(
			"(lower(name) LIKE '%%' || $%d || '%%' OR lower(domain) = $%d OR lower(COALESCE(area,'')) LIKE '%%' || $%d || '%%')",
			i+1, i+1, i+1,
		))
		args = append(args, w)
	}

	if len(conditions) == 0 {
		return nil, nil
	}

	query := fmt.Sprintf(`
		SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
		FROM catalog_items
		WHERE enabled = true AND (%s)
		LIMIT 20
	`, strings.Join(conditions, " OR "))

	rows, err := r.db.Query(query, args...)
	if err != nil {
		return nil, fmt.Errorf("text search: %w", err)
	}
	defer rows.Close()

	for rows.Next() {
		var c Candidate
		rows.Scan(&c.ItemID, &c.Kind, &c.Domain, &c.EntityID, &c.Name, &c.Area)
		// Score based on word overlap
		nameLower := strings.ToLower(c.Name)
		matchCount := 0
		for _, w := range words {
			if strings.Contains(nameLower, w) || strings.ToLower(c.Domain) == w {
				matchCount++
			}
		}
		c.Score = 0.4 + 0.15*float64(matchCount)
		candidates = append(candidates, c)
	}

	return candidates, nil
}

// extractKeywords pulls meaningful words from user text.
func extractKeywords(text string) []string {
	stopWords := map[string]bool{
		"the": true, "a": true, "an": true, "to": true, "in": true,
		"on": true, "off": true, "turn": true, "set": true, "make": true,
		"please": true, "can": true, "you": true, "my": true, "me": true,
		"and": true, "or": true, "but": true, "is": true, "it": true,
		"of": true, "for": true, "with": true, "at": true, "from": true,
		"up": true, "down": true, "do": true, "this": true, "that": true,
		"what": true, "how": true, "all": true, "i": true, "hey": true,
	}

	words := strings.Fields(strings.ToLower(text))
	var keywords []string
	for _, w := range words {
		// Strip punctuation
		w = strings.TrimFunc(w, func(r rune) bool {
			return r < 'a' || r > 'z'
		})
		if len(w) > 0 && !stopWords[w] {
			keywords = append(keywords, w)
		}
	}
	return keywords
}

func dedupeCandidates(candidates []Candidate) []Candidate {
	if len(candidates) < 2 {
		return candidates
	}
	seen := make(map[string]Candidate, len(candidates))
	order := make([]string, 0, len(candidates))
	for _, c := range candidates {
		if current, ok := seen[c.EntityID]; ok {
			if c.Score > current.Score {
				seen[c.EntityID] = c
			}
			continue
		}
		order = append(order, c.EntityID)
		seen[c.EntityID] = c
	}

	out := make([]Candidate, 0, len(seen))
	for _, id := range order {
		if c, ok := seen[id]; ok {
			out = append(out, c)
		}
	}
	return out
}

File: orchestrator-rs/src/resolver.rs
use crate::ml_clients::EmbedderClient;
use crate::models::Candidate;
use sqlx::postgres::PgPoolOptions;
use sqlx::{PgPool, Row};
use std::collections::{HashMap, HashSet};
use std::sync::Arc;

pub struct Resolver {
    db: PgPool,
    embedder: Option<Arc<EmbedderClient>>,
}

impl Resolver {
    pub async fn new(dsn: &str, embedder: Option<Arc<EmbedderClient>>) -> Result<Self, sqlx::Error> {
        let pool = PgPoolOptions::new()
            .max_connections(5)
            .min_connections(2)
            .connect(dsn)
            .await?;
            
        Ok(Self { db: pool, embedder })
    }

    pub async fn ping(&self) -> Result<(), sqlx::Error> {
        sqlx::query("SELECT 1").execute(&self.db).await?;
        Ok(())
    }

    pub async fn resolve(&self, text: &str) -> Result<Vec<Candidate>, sqlx::Error> {
        let lower = text.trim().to_lowercase();

        // Pass A: Lexical matching
        let mut lexical = self.lexical_match(&lower).await.unwrap_or_default();
        lexical = Self::dedupe_candidates(lexical);

        // Strong match short-circuit
        for c in &lexical {
            if c.score >= 0.95 {
                tracing::info!("Strong lexical match: {} ({:.2})", c.entity_id, c.score);
                return Ok(vec![c.clone()]);
            }
        }

        // Pass B: Vector Search
        let mut vector = Vec::new();
        if let Some(embedder) = &self.embedder {
            if let Ok(emb) = embedder.embed(&lower).await {
                vector = self.vector_search(&emb).await.unwrap_or_default();
            } else {
                tracing::warn!("Embedder error, falling back to text search");
            }
        }

        if vector.is_empty() {
            vector = self.text_search(&lower).await.unwrap_or_default();
        }
        vector = Self::dedupe_candidates(vector);

        // Merge
        let mut by_id = HashMap::with_capacity(lexical.len() + vector.len());
        let mut order = Vec::with_capacity(lexical.len() + vector.len());

        for c in lexical {
            if !by_id.contains_key(&c.entity_id) {
                order.push(c.entity_id.clone());
            }
            by_id.insert(c.entity_id.clone(), c);
        }

        for c in vector {
            if let Some(existing) = by_id.get_mut(&c.entity_id) {
                if c.score > existing.score {
                    *existing = c;
                }
            } else {
                order.push(c.entity_id.clone());
                by_id.insert(c.entity_id.clone(), c);
            }
        }

        let mut merged: Vec<Candidate> = order.into_iter().filter_map(|id| by_id.remove(&id)).collect();

        // Sort descending
        merged.sort_by(|a, b| b.score.partial_cmp(&a.score).unwrap_or(std::cmp::Ordering::Equal));

        if merged.len() > 8 {
            merged.truncate(8);
        }

        Ok(merged)
    }

    async fn vector_search(&self, embedding: &[f64]) -> Result<Vec<Candidate>, sqlx::Error> {
        if embedding.is_empty() {
            return Ok(Vec::new());
        }

        // pgvector requires f32
        let f32_embedding: Vec<f32> = embedding.iter().map(|&x| x as f32).collect();
        let vector = pgvector::Vector::from(f32_embedding);

        let query = r#"
            SELECT item_id, kind, domain, entity_id, name, COALESCE(area, ''),
                   1 - (embedding <=> $1) AS score
            FROM catalog_items
            WHERE enabled = true
            ORDER BY embedding <=> $1
            LIMIT 8
        "#;

        let rows = sqlx::query(query)
            .bind(vector)
            .fetch_all(&self.db)
            .await?;

        let mut candidates = Vec::with_capacity(rows.len());
        for row in rows {
            candidates.push(Candidate {
                item_id: row.try_get::<i64, _>(0)?,
                kind: row.try_get::<String, _>(1)?,
                domain: row.try_get::<String, _>(2)?,
                entity_id: row.try_get::<String, _>(3)?,
                name: row.try_get::<String, _>(4)?,
                area: row.try_get::<String, _>(5)?,
                score: row.try_get::<f64, _>(6)?,
            });
        }

        Ok(candidates)
    }

    async fn lexical_match(&self, text: &str) -> Result<Vec<Candidate>, sqlx::Error> {
        let mut candidates = Vec::new();

        // Exact entity_id match
        let query_exact = r#"
            SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
            FROM catalog_items
            WHERE enabled = true AND entity_id = $1
        "#;
        
        let exact_rows = sqlx::query(query_exact)
            .bind(text)
            .fetch_all(&self.db)
            .await?;

        for row in exact_rows {
            candidates.push(Candidate {
                item_id: row.try_get(0)?,
                kind: row.try_get(1)?,
                domain: row.try_get(2)?,
                entity_id: row.try_get(3)?,
                name: row.try_get(4)?,
                area: row.try_get(5)?,
                score: 1.0,
            });
        }

        // Name match
        let words = Self::extract_keywords(text);
        if !words.is_empty() {
            for word in words {
                if word.len() < 3 {
                    continue;
                }
                
                let query_name = r#"
                    SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
                    FROM catalog_items
                    WHERE enabled = true AND lower(name) LIKE '%' || $1 || '%'
                    LIMIT 10
                "#;
                
                let name_rows = sqlx::query(query_name)
                    .bind(&word)
                    .fetch_all(&self.db)
                    .await;
                
                if let Ok(rows) = name_rows {
                    for row in rows {
                        let name: String = row.try_get(4)?;
                        let name_lower = name.to_lowercase();
                        
                        let score = if name_lower == text {
                            0.95
                        } else if name_lower.contains(text) {
                            0.85
                        } else {
                            0.6 + 0.1 * (word.len() as f64) / (name_lower.len() as f64)
                        };

                        candidates.push(Candidate {
                            item_id: row.try_get(0)?,
                            kind: row.try_get(1)?,
                            domain: row.try_get(2)?,
                            entity_id: row.try_get(3)?,
                            name,
                            area: row.try_get(5)?,
                            score,
                        });
                    }
                }
            }
        }

        Ok(candidates)
    }

    async fn text_search(&self, text: &str) -> Result<Vec<Candidate>, sqlx::Error> {
        let words = Self::extract_keywords(text);
        if words.is_empty() {
            return Ok(Vec::new());
        }

        let mut conditions = Vec::with_capacity(words.len());
        for i in 0..words.len() {
            let idx = i + 1;
            conditions.push(format!(
                "(lower(name) LIKE '%' || ${} || '%' OR lower(domain) = ${} OR lower(COALESCE(area,'')) LIKE '%' || ${} || '%')",
                idx, idx, idx
            ));
        }

        let query_str = format!(
            r#"
            SELECT item_id, kind, domain, entity_id, name, COALESCE(area, '')
            FROM catalog_items
            WHERE enabled = true AND ({})
            LIMIT 20
            "#, 
            conditions.join(" OR ")
        );

        let mut query = sqlx::query(&query_str);
        for w in &words {
            query = query.bind(w);
        }

        let rows = query.fetch_all(&self.db).await?;
        let mut candidates = Vec::with_capacity(rows.len());

        for row in rows {
            let name: String = row.try_get(4)?;
            let domain: String = row.try_get(2)?;
            let name_lower = name.to_lowercase();
            let domain_lower = domain.to_lowercase();
            
            let mut match_count = 0;
            for w in &words {
                if name_lower.contains(w) || domain_lower == *w {
                    match_count += 1;
                }
            }

            candidates.push(Candidate {
                item_id: row.try_get(0)?,
                kind: row.try_get(1)?,
                domain,
                entity_id: row.try_get(3)?,
                name,
                area: row.try_get(5)?,
                score: 0.4 + 0.15 * (match_count as f64),
            });
        }

        Ok(candidates)
    }

    fn extract_keywords(text: &str) -> Vec<String> {
        let set: HashSet<&'static str> = vec![
            "the", "a", "an", "to", "in", "on", "off", "turn", "set", "make",
            "please", "can", "you", "my", "me", "and", "or", "but", "is", "it",
            "of", "for", "with", "at", "from", "up", "down", "do", "this", "that",
            "what", "how", "all", "i", "hey",
        ].into_iter().collect();

        text.split_whitespace()
            .map(|w| {
                w.chars()
                    .filter(|c| c.is_ascii_lowercase())
                    .collect::<String>()
            })
            .filter(|w| !w.is_empty() && !set.contains(w.as_str()))
            .collect()
    }

    fn dedupe_candidates(candidates: Vec<Candidate>) -> Vec<Candidate> {
        if candidates.len() < 2 {
            return candidates;
        }

        let mut seen: HashMap<String, Candidate> = HashMap::with_capacity(candidates.len());
        let mut order: Vec<String> = Vec::with_capacity(candidates.len());

        for c in candidates {
            if let Some(existing) = seen.get_mut(&c.entity_id) {
                if c.score > existing.score {
                    *existing = c;
                }
            } else {
                order.push(c.entity_id.clone());
                seen.insert(c.entity_id.clone(), c);
            }
        }

        order.into_iter().filter_map(|id| seen.remove(&id)).collect()
    }
}
